source/extensions/filters/http/custom_response/custom_response_filter.cc

```diff
@@ -43,15 +44,25 @@ Http::FilterHeadersStatus CustomResponseFilter::encodeHeaders(Http::ResponseHead
     return filter_state->policy->encodeHeaders(headers, end_stream, *this);
   }
 
-  // Check for route specific config.
-  const FilterConfig* per_route_settings =
-      Http::Utility::resolveMostSpecificPerFilterConfig<FilterConfig>(decoder_callbacks_);
-  const FilterConfig* config_to_use = per_route_settings
-                                          ? static_cast<const FilterConfig*>(per_route_settings)
-                                          : static_cast<const FilterConfig*>(config_.get());
-  // Check if any custom response policy applies to this response.
-  const PolicySharedPtr policy =
-      config_to_use->getPolicy(headers, encoder_callbacks_->streamInfo());
+  // Traverse up route typed per filter hierarchy till we find a matching
+  // policy. Note that since the traversal is least to most specific, we can't
+  // return early when a match is found.
+  PolicySharedPtr policy;
+  decoder_callbacks_->traversePerFilterConfig(
+      [&policy, &headers, this](const Router::RouteSpecificFilterConfig& config) {
+        const FilterConfig* typed_config = dynamic_cast<const FilterConfig*>(&config);
+        if (typed_config) {
+          // Check if a match is found first to avoid overwriting policy with an
+          // empty shared_ptr.
+          auto maybe_policy = typed_config->getPolicy(headers, encoder_callbacks_->streamInfo());
```

Comment:
IIUC this call may change the `headers`, so depending on the config, one may see different results.
Will it be possible to keep a reference/pointer to the most-specific FilterConfig, and execute `getPolicy()` only once?

Reply:
I don't follow your question Adi.

Each typed config has a `Matcher` that `getPolicy` calls `evaluateMatch` on using the response header to find a match. The intent is to find use a matching policy from the most specific typed config. 

Because the traversal is from least to most specific, we can't stop when a match is found. We want a match in a more specific config to overwrite the one in a less specific config.

I modified `getPolicy`'s interface to accept `const` headers. (I missed this before).